element_facemap/facemap_train.py

```diff
@@ -0,0 +1,447 @@
+            # Fetch model file attachment so that model_file (.pth) is availible in Path.cwd()
+            model_file = (
+                facemap_inference.FacemapModel.File & {"model_id": retrain_model_id}
+            ).fetch1("model_file")
+
+            # Set train_model object to load preexisting model
+            train_model.model_name = model_file
+
+            # Overwrite default train_model.net
+            train_model.net.load_state_dict(
+                torch.load(model_file, map_location=train_model.device)
+            )
+
+            # link model to torch device
+            train_model.net.to(train_model.device)
+
+        # Convert videos to images for train input
+        pre_selected_frame_ind = (FacemapModelTrainingTask & key).fetch1(
+            "selected_frame_ind"
+        )
+
+        # Currently, only support single video training
+        assert len(video_files) == 1
+        video_file = video_files[0]
+
+        # Load video capture to iterate through frames and convert to grayscale
+        cap = cv2.VideoCapture(video_file)
+        if len(pre_selected_frame_ind) == 0:  # set selected frames to all frames
+            selected_frame_indices = np.arange(int(cap.get(cv2.CAP_PROP_FRAME_COUNT)))
+        else:
+            selected_frame_indices = pre_selected_frame_ind
+        frames = []
+        for frame_ind in selected_frame_indices:
+            if int(cap.get(cv2.CAP_PROP_POS_FRAMES)) != frame_ind:
+                cap.set(cv2.CAP_PROP_POS_FRAMES, frame_ind)
+            ret, frame = cap.read()
+            gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+            if ret:
+                frames.append(gray_frame)
+            else:
+                print("Error reading frame")
+        image_data = np.array(frames)
+
+        # Load image frames from video
+        # image_data = utils.load_images_from_video(video_file, selected_frame_ind)
+
+        keypoints_data = utils.load_keypoints(
+            list(zip(*facemap_inference.BodyPart.contents))[0], keypoints_file
+        )
+
+        # Model Parameters (fetch from TrainingParamSet as dict)
+        training_params = (
+            FacemapTrainParamSet & f'paramset_idx={key["paramset_idx"]}'
+        ).fetch1("params")
+        refined_model_prefix = (FacemapModelTrainingTask & key).fetch1(
+            "refined_model_prefix"
+        )  # default = "refined_model"
+
+        # Train model using train function defined in Pose class
+        train_model.train(
+            image_data,
+            keypoints_data.T,  # needs to be transposed
+            int(training_params["epochs"]),
+            int(training_params["batch_size"]),
+            float(training_params["learning_rate"]),
+            int(training_params["weight_decay"]),
+            bbox=training_params["bbox"],
+        )
+
+        # Save Refined Model
+        model_output_path = output_dir / f"{refined_model_prefix}.pth"
```

Comment:
This looks like the `refined_model_prefix` is used as the model name, not exactly as a "prefix". Is this intended? or you mean to prepend this prefix with the name of the "retrain_model"?

Reply:
Ah yes, I've updated it to prepend "retrain_model".